enzyme/Enzyme/GradientUtils.cpp

```diff
@@ -3301,88 +3438,114 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
       Value *b = getNewFromOriginal(arg->getOperand(1 + i));
       invertargs.push_back(b);
     }
+    Value *ip = invertPointerM(arg->getPointerOperand(), bb);
+
+    auto rule = [&](Value *ip) {
 #if LLVM_VERSION_MAJOR > 7
-    auto shadow = bb.CreateGEP(
-        cast<PointerType>(arg->getPointerOperandType())->getElementType(),
-        invertPointerM(arg->getPointerOperand(), bb), invertargs,
-        arg->getName() + "'ipg");
+      auto shadow =
+          bb.CreateGEP(cast<PointerType>(ip->getType())->getElementType(), ip,
+                       invertargs, arg->getName() + "'ipg");
 #else
-    auto shadow = bb.CreateGEP(invertPointerM(arg->getPointerOperand(), bb),
-                               invertargs, arg->getName() + "'ipg");
+      auto shadow = bb.CreateGEP(ip, invertargs, arg->getName() + "'ipg");
 #endif
-    if (auto gep = dyn_cast<GetElementPtrInst>(shadow))
-      gep->setIsInBounds(arg->isInBounds());
+
+      if (auto gep = dyn_cast<GetElementPtrInst>(shadow))
+        gep->setIsInBounds(arg->isInBounds());
+
+      return shadow;
+    };
+
+    Value *shadow = applyChainRule(arg->getType(), bb, rule, ip);
+
     invertedPointers.insert(
         std::make_pair((const Value *)oval, InvertedPointerVH(this, shadow)));
     return shadow;
   } else if (auto inst = dyn_cast<AllocaInst>(oval)) {
     IRBuilder<> bb(getNewFromOriginal(inst));
     Value *asize = getNewFromOriginal(inst->getArraySize());
-    AllocaInst *antialloca = bb.CreateAlloca(
-        inst->getAllocatedType(), inst->getType()->getPointerAddressSpace(),
-        asize, inst->getName() + "'ipa");
-    invertedPointers.insert(std::make_pair(
-        (const Value *)oval, InvertedPointerVH(this, antialloca)));
-    if (inst->getAlignment()) {
+
+    auto rule1 = [&]() {
```

Comment:
Similar allocation comment here.

Reply:
Since we cannot judge the vectorisation of coalesced allocas right now, i would go with the most canonical form and keep it this way.